semantic/poly_types.go

```diff
@@ -85,22 +93,36 @@ func (tv Tvar) freeVars(c *Constraints) TvarSet {
 	}
 	return fvs
 }
-func (l Tvar) unifyType(kinds map[Tvar]Kind, r PolyType) (Substitution, error) {
-	switch r := r.(type) {
+func (l Tvar) unifyType(kinds map[Tvar]Kind, s Substitution, r PolyType) (Substitution, error) {
+	// Use the substitution to reduce the left and right hand sides.
+	// This ensures we never add a transitive cycle in the substitution.
+	tl, ok := l.apply(s)
+	for ok {
+		tl, ok = tl.apply(s)
+	}
+	tr, ok := r.apply(s)
+	for ok {
+		tr, ok = tr.apply(s)
+	}
```

Comment:
I have some questions and I am going to use this section of code to demonstrate them.

As I understand it the overall change is to no longer merge substitutions together but to grow the substitution and never remove from it. This means that a substitution can contains chains of subs.

For example give a  substitution  like this:

```
t0 -> t1 
```

And another sub like this:

```
t1 -> int
```

Previously we would have merged the subs to become this:

```
t0 -> int
```

And lost the information about `t1`.
Previously the assumption as that it was safe to loose information about `t1` because we would have substituted all instances of `t1` with `int` already? 

This change simply grows the substitution instead:

```
t0 -> t1
t1 -> int
```

Therefore in order to completely apply the substitution we have to re-apply it until nothing changes. I.e a fixed-point algorithm.


Some of my questions:

1. How does this change actually solve the issue? Was it because the assumption the it was safe to loose information in the substitution was violated? If so can we track down were that happened?
2. Why is the loop part of this code and not part of the substitution type itself? In other words, how do I know i am supposed to apply in a loop here but not in other places?
3. If `s`, `l`, and `r` are all inputs to this function why was the substitution not applied before passing them into this function? In other words, how do I know its my responsibility to apply the substitution here in addition to performing the unification.

Reply:
The example you give is not quite related to this issue. The issue was that you could have the same Tvar bound in both substitutions. `t0 -> t1` in one, and `t0 -> int` in the other. In this case, you must unify `t1` and `int`, but that never happened, so you never got `t1 -> int` in the returned substitution.

Having said that and going back to your example, it is not safe to lose information about a Tvar in a substitution unless you also perform the same substitution over the remaining constraints. 

Addressing point 2, yes I can add a method to the substitution so that the loop is hidden and that applying a substitution to a type automatically reduces the type completely. Because you should always apply a substitution that way.

As for point 3, only in Tvar unification can a transitive cycle be added to the substitution. This is why you **must** use the substitution to reduce both sides of the current constraint before adding it to the substitution. This is the **only** place where this occurs.